amr-wind/wind_energy/ABLBoundaryPlane.cpp

```diff
@@ -0,0 +1,624 @@
+    m_tnp1 = times[idxp1];
+    AMREX_ALWAYS_ASSERT(((m_tn <= time) and (time <= m_tnp1)));
+
+    const int normal = ori.coordDir();
+    const amrex::GpuArray<int, 2> perp = get_perpendicular_idx(normal);
+
+    const auto& bx = (*m_data_n[ori])[lev].box();
+    const auto& lo = bx.loVect();
+    const size_t n0 = bx.length(perp[0]);
+    const size_t n1 = bx.length(perp[1]);
+
+    amrex::Vector<size_t> start{static_cast<size_t>(idx),
+                                static_cast<size_t>(lo[perp[0]]),
+                                static_cast<size_t>(lo[perp[1]]), 0};
+    amrex::Vector<size_t> count{1, n0, n1, nc};
+    amrex::Vector<amrex::Real> buffer(n0 * n1 * nc);
+    grp.var(fld->name()).get(buffer.data(), start, count);
+
+    const auto& datn = ((*m_data_n[ori])[lev]).array();
+    auto d_buffer = buffer.dataPtr();
+    amrex::LoopOnCpu(
+        bx, nc, [=] AMREX_GPU_DEVICE(int i, int j, int k, int n) noexcept {
+            const int i0 = plane_idx(i, j, k, perp[0], lo[perp[0]]);
+            const int i1 = plane_idx(i, j, k, perp[1], lo[perp[1]]);
+            datn(i, j, k, n + nstart) = d_buffer[((i0 * n1) + i1) * nc + n];
+        });
+
+    start[0] = static_cast<size_t>(idxp1);
+    grp.var(fld->name()).get(buffer.data(), start, count);
+
+    const auto& datnp1 = ((*m_data_np1[ori])[lev]).array();
+    amrex::LoopOnCpu(
+        bx, nc, [=] AMREX_GPU_DEVICE(int i, int j, int k, int n) noexcept {
+            const int i0 = plane_idx(i, j, k, perp[0], lo[perp[0]]);
+            const int i1 = plane_idx(i, j, k, perp[1], lo[perp[1]]);
+            datnp1(i, j, k, n + nstart) = d_buffer[((i0 * n1) + i1) * nc + n];
+        });
+
+    ((*m_data_n[ori])[lev]).prefetchToDevice();
+    ((*m_data_np1[ori])[lev]).prefetchToDevice();
+}
+
+void InletData::interpolate(const amrex::Real time)
+{
+    m_ti = time;
+    for (amrex::OrientationIter oit; oit; ++oit) {
+        auto ori = oit();
+        if (not this->is_populated(ori)) continue;
+
+        const int nlevels = m_data_n[ori]->size();
+        for (int lev = 0; lev < nlevels; ++lev) {
+
+            const auto& datn = (*m_data_n[ori])[lev];
+            const auto& datnp1 = (*m_data_np1[ori])[lev];
+            auto& dati = (*m_data_interp[ori])[lev];
+
+            dati.linInterp<amrex::RunOn::Device>(
+                datn, 0, datnp1, 0, m_tn, m_tnp1, m_ti, dati.box(), 0,
+                dati.nComp());
+        }
+    }
+}
+
+bool InletData::is_populated(amrex::Orientation ori) const
+{
+    return m_data_n[ori] != nullptr;
+}
+
+ABLBoundaryPlane::ABLBoundaryPlane(CFDSim& sim)
+    : m_time(sim.time()), m_repo(sim.repo()), m_mesh(sim.mesh())
+{
+    amrex::ParmParse pp("ABL");
+    pp.query("bndry_write_frequency", m_write_frequency);
+    pp.queryarr("bndry_output_planes", m_out_planes);
+    pp.query("bndry_output_start_time", m_out_start_time);
+    pp.queryarr("bndry_output_var_names", m_out_var_names);
+    pp.queryarr("bndry_input_var_names", m_in_var_names);
```

Comment:
Did you have a specific use-case that would require different sets of input/output variables? Can this be collapsed into one?

Reply:
no use case. Collapsed into one.